cosmos/rpc.go

```diff
@@ -91,23 +91,43 @@ func (c *RPC) QueryWithData(path string, data []byte) ([]byte, int64, error) {
 
 // BuildAndBroadcastMsg builds and signs message and broadcast it to node.
 func (c *RPC) BuildAndBroadcastMsg(msg sdktypes.Msg) (*abci.ResponseDeliverTx, error) {
-	c.broadcastMutex.Lock() // Lock the whole signature + broadcast of the transaction
-	signedTx, err := c.CreateAndSignTx([]sdktypes.Msg{msg})
+	signedTx, err := c.buildAndBroadcastMsgNoResult(msg)
 	if err != nil {
-		c.broadcastMutex.Unlock()
 		return nil, err
 	}
+	return c.waitForTxResult(signedTx)
+}
 
-	txres, err := c.BroadcastTxSync(signedTx)
-	c.broadcastMutex.Unlock()
+func (c *RPC) buildAndBroadcastMsgNoResult(msg sdktypes.Msg) (tenderminttypes.Tx, error) {
```

Comment:
I split this function into 2 to be able to use a defer on the mutex unlock instead of putting it in errors.

Reply:
you could as well explicitly unlock the mutex without a defer but I'm good with that